Peak overlay in Instrument View

jclarkeSTFC opened the pull request:
Peak overlay for new Instrument View. Peaks workspaces with the same instrument as the main workspace will show on the left of the interface, and can overlaid by selecting their checkbox. If a peaks workspace is overlaid and a detector is selected in the projection with a peak, then that peak will also be overlaid on the 1D plot with the HKL value.

Closes #40043  <!-- One line per closed issue. -->

<!-- If issue raised by user. Do not leak email addresses.
**Report to:** [user name]
-->

### To test:

To aid testing, you can have both the old and new instrument view interfaces open at the same time, and pick/delete peaks in the old one. The new one should update.

You can't pick peaks in the new interface yet.

<!-- Include sufficient instructions for someone unfamiliar with the application to test.
Ok to refer back to instructions in the issue.
-->

<!-- REMEMBER:
- Add labels, milestones, etc.
- Ensure the base of this PR is correct (e.g. release-next or main)
- Add release notes in separate file as per ([guidelines](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)), or justify their absence:
  *This does not require release notes* because <fill in an explanation of why>
-->

<!--  GATEKEEPER: When squashing, remove the section from HERE...  -->
---

### Reviewer

**Your comments will be used as part of the gatekeeper process.** Comment clearly on what you have checked and tested during your review. Provide an audit trail for any changes requested.

As per the [review guidelines](http://developer.mantidproject.org/ReviewingAPullRequest.html):

- Is the code of an acceptable quality? ([Code standards](http://developer.mantidproject.org/Standards/)/[GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html))
- Has a thorough functional test been performed? Do the changes handle unexpected input/situations?
- Are appropriately scoped unit and/or system tests provided?
- Do the release notes conform to the [guidelines](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html) and describe the changes appropriately?
- Has the relevant (user and developer) documentation been added/updated?

### Gatekeeper

As per the [gatekeeping guidelines](https://developer.mantidproject.org/Gatekeeping.html):

- Has a thorough first line review been conducted, including functional testing?
- At a high-level, is the code quality sufficient?
- Are the base, milestone and labels correct?
<!--  GATEKEEPER: ...To HERE  -->

jclarkeSTFC committed:
Plot peaks from selected peaks workspaces

jclarkeSTFC committed:
Overlay peaks on 1D plot if detector selected

jclarkeSTFC committed:
Monitor changes to peaks workspaces and update

jclarkeSTFC committed:
Make Peak a dataclass

jclarkeSTFC committed:
Added tests for peak overlay

jclarkeSTFC committed:
Added release note.

jclarkeSTFC committed:
Fixes from review

GuiMacielPereira reviewed (dismissed):
Tested it by creating peaks with the old instrument view and confirming they appear in the new one. Deleting rows of the peak workspace also updates the new instrument view correctly. The unit tests are also extensive.

RichardWaiteSTFC reviewed (changes requested):
Haven't had a look through the code but here are some first impressions after testing (which are largely positive!):

1. Often it is necessary to overlay multiple peak tables - this is possible here but the color of the markers is the same, could they be different? Ideally one could even choose the color (but that can be done later)
2. I really like the vertical lines on the 1D plot (a lot clearer than the old UI). It works very nicely for plotting different spectra as well, but I wonder if the color of the peak line could match that of the data? 
3. Could the HKL labels be at the top rather than bottom of the 1D plot axes - it clashes with the x-axis label atm.
4. I like the check box list of peak workspaces (especially nice it doesn't include tables from other instruments), it makes it clear which table is plotted. It would also be nice to be able to drag and drop peak tables onto the UI as well though if it wasn't too difficult? One could have many tables with similar long names, but the workspace list in the main GUI can be sorted by last modified to make it easier to find the right table.
5. Could the markers be slightly smaller/transparent to not obscure the data beneath - or maybe a cross rather than filled-in marker?
6. The peaks plotted need to respect the integration limits of the colorfill plot
7. Could the peak table list update when the ADS is updated - so that new peak tables are available to overlay?
8. I like the peak labels, but where there are multiple peaks in a single detector (and there can be or order ~10!) the old instrument view would plot text `HKL` or similar. This was unhelpful in that I couldn't see what direction in HKL that corresponded to, but it did look cleaner. Would it be possible to represent e.g. the (1,1,0), (2,2,0) set of peaks with a single label but make it clear in some way there are other peaks? Maybe `[1,1,0] x N`? Note square brackets because in this case it represents a direction (crystallographers are very particular about brackets...)

jclarkeSTFC committed:
Colour each peaks workspace differently

jclarkeSTFC committed:
Plot 1D HKL labels above plot instead of below

jclarkeSTFC committed:
Drag and drop peaks workspaces to select them

jclarkeSTFC committed:
Improve peak HKL labels

Use correct bracket, changed font to times to make the different
brackets visible, and made the label more concise when a detector has
multiple peaks

jclarkeSTFC committed:
Synchronise peak colours in projection and 1D

The text in the peaks workspaces list will also be the same colour, to
act as a legend

jclarkeSTFC committed:
Update peaks workspaces on add to ADS

When new workspaces are added to the ADS then the peaks workspaces will
update

RichardWaiteSTFC reviewed (changes requested):
Thanks for this - really like the 1D plot lines being the same color as the peak marker! And the drag and drop is nice as well! All seems to be working well, peaks are updated with ADS replace etc.

Just a few things I've noticed:
a) Think `tab:blue` color (first color in peak marker cycle) isn't very compatible with the detector/data colormap - it's very difficult to see the marker.
b) Would it be possible to have peak HKL labels the same color as the marker
c) I can no longer see the HKL labels in the 1D plot - is that deliberate, or something to do with my window size/resolution?
d) Sometimes we use different UBs on different tables, so a HKL label might be different for a given peak - here it's hard to see which is which. The different colors form point (b) would help, maybe a different offset (alternate between left/right, top/bottom translation)
<img width="269" height="175" alt="image" src="https://github.com/user-attachments/assets/92ccb8f8-9c13-496b-8f8e-a0518b16fad6" />
e) It's also quite difficult to see the data under the peak marker - can the peak marker be smaller or maybe a cross which could be larger without obscuring data? Or at least use a transparent marker face?

I'd like c to be fixed (if indeed it is broken) in this PR if at all possible, but happy for the others to wait. Feel free to push back on any of the above!

jclarkeSTFC committed:
Improvements to peak label visibility

jclarkeSTFC committed:
Redraw peaks on projection change

jclarkeSTFC committed:
Fix peak locations in certain projections

We must get the peak projected positions from the original calculation,
because the projected detector positions are adjusted based on the
other points, it's not a calculation done in isolation. So now we cache
the detector positions on the model.

RichardWaiteSTFC reviewed (changes requested):
Everything looks great, just one problem, I think sometimes the 1D plot shows a different HKL label? Or there's some sort of mix up!

This pixel was meant to have 4 peaks apparently - and the HKL is different form the one indicated on the 2D plot

<img width="616" height="829" alt="image" src="https://github.com/user-attachments/assets/c4dc9b4d-e196-429f-8094-144a71750072" />

jclarkeSTFC committed:
Allow masked workspaces

If a detector is masked, then don't show the peaks on it. If there are
no visible peaks after masking, then just don't show any points.

jclarkeSTFC committed:
Fix opeak overlay in 3D

Need to reset the projection position cache when changing to 3D view.

jclarkeSTFC committed:
Look up detector ID indices in a different way

By using np.where(np.isin) the order of the original objects is not
preserved, so the labels were then wrong. This code maintains the order
of the original objects. The sort + search is for speed, because then
the search can use a binary search (fast).

RichardWaiteSTFC reviewed (changes requested):
Looks great - thanks for fixing the labels, everything matches up now!
Just found a couple more minor things sorry!

1. Really minor, but when there are multiple peaks would it be possible to pick the peak at the highest d-spacing - i.e. in this case the (-2,-2,-1) instead of  (-4, -4, -2)?
<img width="399" height="737" alt="image" src="https://github.com/user-attachments/assets/42a5c2c3-ab1e-4ff3-9c9a-d9ff648ea916" />
2. Peak overlay doesn't seem to work on the 1D plot when the units are momentum transfer?

Neither of these is a huge deal so can do in another PR?

jclarkeSTFC committed:
Use the peak with highest d-spacing for the label

jclarkeSTFC committed:
Overlay in wavelength and q as well

jclarkeSTFC committed:
Fix tests following change to Peak constructor

RichardWaiteSTFC reviewed (approved):
Looks and works great - thanks for this!
